sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -514,38 +443,26 @@ case class GpuHashAggregateExec(
   }
 
   /**
-   * concatenateBatches - given two ColumnarBatch instances, return a sequence of GpuColumnVector
-   * that is the concatenated columns of the two input batches.
-   * @param aggregatedInputCb this is an incoming batch
-   * @param aggregatedCb this is a batch that was kept for concatenation
-   * @return Seq[GpuColumnVector] with concatenated vectors
+   * Concatenates batches by concatenating the corresponding column vectors within the batches.
+   * @note the input batches are not closed as part of this operation
+   * @param batchesToConcat batches to concatenate
+   * @return concatenated vectors that together represent the concatenated batch result
    */
-  private def concatenateBatches(aggregatedInputCb: ColumnarBatch,
-      aggregatedCb: ColumnarBatch,
-      concatTime: GpuMetric): Seq[GpuColumnVector] = {
+  private def concatenateBatches(
+      batchesToConcat: mutable.ArrayBuffer[LazySpillableColumnarBatch]): Seq[GpuColumnVector] = {
+    val concatTime = metrics.concatTime
     withResource(new NvtxWithMetrics("concatenateBatches", NvtxColor.BLUE, concatTime)) { _ =>
-      // get tuples of columns to concatenate
-
-      val zipped = (0 until aggregatedCb.numCols()).map { i =>
-        (aggregatedInputCb.column(i), aggregatedCb.column(i))
-      }
-
-      zipped.map {
-        case (col1, col2) =>
-          GpuColumnVector.from(
-            cudf.ColumnVector.concatenate(
-              col1.asInstanceOf[GpuColumnVector].getBase,
-              col2.asInstanceOf[GpuColumnVector].getBase), col1.dataType())
+      val numCols = batchesToConcat.head.numCols
+      (0 until numCols).safeMap { i =>
+        val columnType = batchesToConcat.head.getBatch.column(i).dataType()
+        val columnsToConcat = batchesToConcat.map {
+          _.getBatch.column(i).asInstanceOf[GpuColumnVector].getBase
+        }
+        GpuColumnVector.from(cudf.ColumnVector.concatenate(columnsToConcat: _*), columnType)
```

Comment:
I was running some tests

```
scala> spark.conf.set("spark.sql.shuffle.partitions", "1")

scala> val df = spark.range(4000000000L)
df: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> df.registerTempTable("data")
warning: there was one deprecation warning (since 2.0.0); for details, enable `:setting -deprecation' or `:replay -deprecation'

scala> spark.time(spark.sql("SELECT COUNT(DISTINCT id) from data").show)
```

And I got a leak on this line. There were no other decs or incs to the buffer besides this. I also ran into some cases where I had to drop the batch size to be rather small < 1g to get it to not fail. For me that feels like we have a bug somewhere because the max memory usage should be <= 4x the batch size and this was more than that. I include it here because the leak might be related.

Reply:
Thanks for the example!  I never saw a failure with it, but I was able to reproduce the leak.  The issue is failing to free this when the concatenated result is later merge-aggregated.  I'll post a fix shortly.

Reply:
I started to look at the memory usage of hash aggregate to see if it followed our assumptions about total memory usage.

The size of the memory that the hash table allocates is `2 * number_of_rows * size_of(thrust::pair<size_type, size_type>)` in addition to the original input table, and an output table. I am not 100% sure of the size of the output table, but we might want to ask the cudf team about it.

So in my case where I have a single long as the input with a batch size of ~2GiB and the output size is as big as the input size, then the total GPU memory size is going to approximately be. 2GiB + (2 * 2GiB/sizeof(int64_t) * sizeof(thrust::pair<size_type, size_type>)) + 2GiB => 8 GiB.  This works because the target max size is 4x the batch size.  But if I set the batch size to ~1 GiB I also get 8 GiB as the max memory size, because we are assuming that we can double the input size up to 2 GiB.  And I don't think that is true here.

We might need to rethink how we size joins as well.

Reply:
For a byte column we would end up with 2 GiB input, 16 GiB of hash table and 256 bytes of output or between 18 and 36x the target batch size.
For a short column we would end up with 2 GiB input, 8 GiB of hash table and 54 KiB of output or between 5 and 10x the target batch size.
For an int column we would end up with 2 GiB input, 4 GiB of hash table and up to 2 GiB of output or between 4 and 8x the target batch size.

I think byte is the worst case we could have, even strings would not be as bad because we would still have the offsets column which would make it close to the int size in the worst case. So I think we need to do a quick look at the input schema and guess what the maximum batch size is we want to not go over the 4x target size limit. I think this is an odd enough corner case that it should not impact too many things.